src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
+            ILoggerFactory loggerFactory,
+            DiagnosticEventForwarder forwarder,
+            RulesetsForProjects rulesetsForProjects)
         {
-            _workspace = workspace;
-            _forwarder = forwarder;
             _logger = loggerFactory.CreateLogger<CSharpDiagnosticService>();
+            _providers = providers;
+
+            workspace.WorkspaceChanged += OnWorkspaceChanged;
 
-            _workspace.WorkspaceChanged += OnWorkspaceChanged;
+            _forwarder = forwarder;
+            _workspace = workspace;
+            _rulesetsForProjects = rulesetsForProjects;
+
+            Task.Run(async () =>
+            {
+                while (!workspace.Initialized || workspace.CurrentSolution.Projects.Count() == 0) await Task.Delay(500);
+
+                QueueForAnalysis(workspace.CurrentSolution.Projects);
+                _initializationQueueDoneSource.Cancel();
+                _logger.LogInformation("Solution initialized -> queue all projects for code analysis.");
+            });
+
+            Task.Factory.StartNew(() => Worker(CancellationToken.None), TaskCreationOptions.LongRunning);
         }
 
-        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
+        private async Task Worker(CancellationToken token)
         {
-            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged)
+            while (!token.IsCancellationRequested)
             {
-                var newDocument = changeEvent.NewSolution.GetDocument(changeEvent.DocumentId);
-
-                this.EmitDiagnostics(newDocument.FilePath);
-                foreach (var document in _workspace.GetOpenDocumentIds().Select(x => _workspace.CurrentSolution.GetDocument(x)))
+                try
+                {
+                    var currentWork = GetThrottledWork();
+                    await Task.WhenAll(currentWork.Select(x => Analyze(x.project, x.workReadySource, token)));
+                    await Task.Delay(100, token);
+                }
+                catch (Exception ex)
                 {
-                    this.EmitDiagnostics(document.FilePath);
+                    _logger.LogError($"Analyzer worker failed: {ex}");
                 }
             }
         }
 
-        public void QueueDiagnostics(params string[] documents)
+        private IEnumerable<(Project project, CancellationTokenSource workReadySource)> GetThrottledWork()
         {
-            this.EmitDiagnostics(documents);
-        }
-
-        private void EmitDiagnostics(params string[] documents)
-        {
-            if (_forwarder.IsEnabled)
+            lock (_workQueue)
             {
-                foreach (var document in documents)
-                {
-                    if (!_openDocuments.Contains(document))
-                    {
-                        _openDocuments.Enqueue(document);
-                    }
-                }
+                var currentWork = _workQueue
```

Comment:
sorry, I dont fully understand how all these logic works. but I would recommend to profile and see whether it meets your perf expectataion.